### PR TITLE
Storage: Wait for ZFS device to appear when activating a volume

### PR DESCRIPTION
**Pull request by MusicDin**
When system is highly loaded, the udev rules may take more than half a second to be applied. As a result, the device may still not be visible when trying to retrieve it's disk path.

Instead of giving a udev rules half a second to be applied, wait up to 30 seconds for the device to appear. This fixes the issue where zvol device cannot be located if the rules are not applied in time.

### CODE REVIEW

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 1]
```diff
@@ -2102,8 +2107,15 @@ func (d *zfs) activateVolume(vol Volume) (bool, error) {
...
 
-		// Wait half a second to give udev a chance to kick in.
-		time.Sleep(500 * time.Millisecond)
+		// Wait up to 30 seconds for the device to appear.
+		ctx, cancel := context.WithTimeout(d.state.ShutdownCtx, 30*time.Second)
+		defer cancel()
+
+		if !tryExists(ctx, filepath.Join("/dev/zvol", dataset)) {
+			// If device has not appeared in time, revert it's volume mode and error out.
+			_ = d.setDatasetProperties(dataset, fmt.Sprintf("volmode=%s", current))
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go:2116]: Can we use a reverter here ?

Comment (MusicDin) [reply lxd/storage/drivers/driver_zfs_volumes.go]: Done

**Review by tomponline — COMMENTED** [review]
So this fixes the issue in your reproducer?

**Comment by MusicDin** [comment]
> So this fixes the issue in your reproducer?

Yes, initializing 100+ containers seems stable. I'm still testing the fix when LXD is clustered

**Comment by MusicDin** [comment]
In clustered LXD I've observed that occasionally this issue happens during deactivation of the volume, where the volume's disk path is obtained in order to ensure the path is removed (disappears).

**Comment by tomponline** [comment]
> The first part is done, so it can be merged. I'm just investigating why deactivation occasionally fails. This can be also a follow up PR if necessary

From @MusicDin

**Review by tomponline — APPROVED** [review]

**Pull request merged**